### PR TITLE
Multilingual Support

### PR DESCRIPTION
**Pull request by JimZhouZZY**
# Multilingual Support Description

## Introduction
This update adds an interface for implementing multilingual functionality to the game. Due to the unfinished translation work, some content is temporarily missing, but it is harmless, the interface has been completed. I will continue to finish Chinese translation.

## How to Add a New Language
1. Create a language directory: Create a copy of the `en` folder under the `data` folder and rename it to the name of the language you need, assuming it is `language_name`. (Due to the unfinished translation work, some content is temporarily missing.)
2. Copy font files: Add the font files required for your language in ttf format under the `fonts` folder.
3. Import font files: Modify the contents of `data/language_name/fonts_description.py` and change the name of the font file to the name of the font file you need.
4. Add language button: Modify the content of the `element_grid` list in the `create_choose_language_menu` function in `menu_creater_manager.py`. Add a button with your language's full name as `title` and `language_name` as `callback`. For example:
   ```python
    element_grid = [
        [Button(title="English", callback=lambda: buttons_callback("en"))],
        [Button(title="中文简体", callback=lambda: buttons_callback("zh_cn"))],
        ...
        [Button(title="Language Full Name", callback=lambda: buttons_callback("language_name"))],
    ]
   ```
5. The content that needs to be translated includes:
    * Strings, function outputs, and dictionaries in `text.py`.
    * The content of each txt file under `data/language_name/maps/level_number`.
    * Some content in each map_properties.tmx file under `data/language_name/maps/level_number`, including:
        + The value of level_name.
        + The value of main_mission_description.
        + The value of limited_mission_description (if it exists).
    * Some content in some xml files under `data/language_name/`.
        + The content within the `<talk>` tag in `characters.xml`.
        + The content within the `<info>` tag in `alternations.xml`.
        + The content within the `<info>` tag in `items.xml`.
        + The content within the `<info>` tag in `skills.xml`.
    Please note that the format of the above changes must be: `<language_name> translated content </language_name>`. For example:
        ```xml
        ...
        <interaction>
            <talk>
                <en> Hurry up ! Leave ... the necropolis. </en>
                <zh_cn> 快呀！从...进入墓地。</zh_cn>
            </talk>
            <talk>
                <en> The clock is ticking... original place. </en>
                <zh_cn> 时钟在滴答...回原来的地方。 </zh_cn>
            </talk>
        </interaction>
        ...
        ```

# Changes and Explanations
1. Created `src/gui/language.py` to handle language-related matters. `language.py` will import the corresponding `text.py` according to the language in `saves/options.xml`.
2. Moved all static strings involving output in all files to variables in `text.py`, with variable names composed of uppercase letters and underscores of English content of strings. If the last character of a variable name is an underscore, it means that there is a colon at the end of the string.
3. Moved all contents involving variables and output in all files to be processed as functions in `text.py`.
4. Corresponded all contents involving class names and output in all files with dictionaries in `text.py`. (The original processing method is temporarily retained for English)
5. Changed the method of reading `<info>` or `<talk>` tags from xml files under the `data` folder. Added `<language_name>` tag. (But still retained the original processing method because translation work has not been completed)
6. Added code related to restarting games in `main.py`. Used to restart games after adjusting languages. Since I couldn't find a better way to restart games, I used creating subprocesses to implement game restart functionality.
7. Changed return value type of variable passed to `quit_game` in each `Scene` to int. And modified contents of function test_exit_game in test_start_screen.py (modification is very small, just changing True to 1)
   + When quit_game == 2 restart game
   + When quit_game == 1 exit game
   + When quit_game == 0 loop continues
8. Created src/gui/info_box.py to rewrite InfoBox class. This is because there is a Close button in InfoBox, and I couldn't find a better way to change its text.
9. Added language-related UI in `Option menu`.

### CODE REVIEW

File: src/gui/language.py
Comment (Grimmys) [inline src/gui/language.py:1]: In my opinion, this is not related to GUI.
It's only loading the user language value, and getting proper translation string from XML element.
It's not like rendering anything to the screen, so I think it should rather be in `src/services`

File: src/gui/info_box.py
Comment (Grimmys) [inline src/gui/info_box.py:1]: I'm not sure why `InfoBox` has been introduce inside of the project instead of keep using the one from `pygamepopup`.

What has been change?
In any way, changes to the components should be done in the appropriate repository, and then, as soon as they are ready, a new version of `pygamepopup` should be released.

Link to the repository: [https://github.com/Grimmys/pygame_popup_manager](https://github.com/Grimmys/pygame_popup_manager)

File: src/services/menu_manager.py
Comment (Grimmys) [inline src/services/menu_manager.py:1]: Same as for `InfoBox` component, changes to these classes have to be done in [https://github.com/Grimmys/pygame_popup_manager](https://github.com/Grimmys/pygame_popup_manager)

File: src/services/scene_manager.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from src.scenes.level_scene import LevelStatus, LevelScene
 from src.scenes.scene import Scene
 from src.scenes.start_scene import StartScene
+from src.gui.language import DATA_PATH
```
Comment (Grimmys) [inline src/services/scene_manager.py:14]: That's the only change in this file, is it required?

File: main.py [hunk 2]
```diff
@@ -26,28 +26,38 @@ def main_loop(
     controller -- the scene manager acting as the main controller of the game
     screen -- the screen on which the current frame rate is displayed
     clock -- the clock regulating the maximum frame rate of the game
+
+    Returns:
+    True -- restart game
+    False -- exit game
     """
-    quit_game: bool = False
+    quit_game: int = False
```
Comment (Grimmys) [inline main.py:34]: This section needs to be improved.

We need to decide if `quit_game` is a `bool` or a `int`, it should not be both at different times.
Also, we should avoid giving unclear signification to arbitrary numbers.

Ideally, it should be using an `Enum`: [https://docs.python.org/3/library/enum.html]( https://docs.python.org/3/library/enum.html).
So then each different value that could be taken by the variable would having a nice label, and also it should not be possible to have value outside of the range of available ones.

Alternatively, constants need to be defined in `src/constants.py`.

File: maps/level_1/map.tmx [hunk 3]
```diff
@@ -1,16 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <map version="1.8" tiledversion="1.8.4" orientation="orthogonal" renderorder="right-down" compressionlevel="0" width="18" height="9" tilewidth="32" tileheight="32" infinite="0" nextlayerid="5" nextobjectid="23">
- <properties>
```
Comment (Grimmys) [inline maps/level_1/map.tmx:3]: I'm not sure about splitting these properties into different files...
Because maps are not generated by hand, they are by using a tool ([Tiled](https://www.mapeditor.org/)).

So it would be adding a burden to this generation to have to split it...

But maybe it's still the best option, let me just think about it a bit.

File: src/game_entities/entity.py [hunk 4]
```diff
@@ -63,10 +64,13 @@ def __str__(self) -> str:
...
-        string_entity: str = self.name.replace("_", " ").title()
-        string_entity = re.sub(
-            r"[0-9]+", "", string_entity
-        )  # Remove numbers like the id
+        try:
+            string_entity: str = dict_entity_names[re.sub(r"[0-9]+", "", self.name).lower()]
+        except KeyError:
+            string_entity: str = self.name.replace("_", " ").title()
+            string_entity = re.sub(
+                r"[0-9]+", "", string_entity
```
Comment (Grimmys) [inline src/game_entities/entity.py:72]: Maybe we should extract this regex somewhere, since it's used twice.

File: src/game_entities/foe.py [hunk 5]
```diff
@@ -135,9 +136,15 @@ def roll_for_loot(self) -> Sequence[Item]:
...
         """Return the list of keywords of the foe in a formatted version"""
-        return ", ".join(
-            [keyword.name.lower().capitalize() for keyword in self.keywords]
-        )
+        formatted_string: str = ""
+        for keyword in self.keywords:
+            try:
+                formatted_string += dict_foe_keywords[keyword.name.lower()] + ", "
+            except KeyError:
+                formatted_string += keyword.name.lower().capitalize()
```
Comment (Grimmys) [inline src/game_entities/foe.py:144]: Huh where is the comma in this case?

**Review by Grimmys — CHANGES_REQUESTED** [review]
Looks good overall, just some small issues

File: src/gui/info_box.py
Comment (JimZhouZZY) [reply src/gui/info_box.py:1]: As mentioned above
>Created src/gui/info_box.py to rewrite InfoBox class. This is because there is a Close button in InfoBox, and I couldn't find a better way to change its text.  

There are two lines changed in total:
One line for import added.
```python
from src.gui.language import STR_CLOSE
```
One line for close button text modified.
```python
Line174 title="Close",
Line174 title=STR_CLOSE,
```

File: src/services/menu_manager.py
Comment (JimZhouZZY) [reply src/services/menu_manager.py:1]: The menu_manager is added bucause it import info_box. 
But I checked it again and found that the import is only related to type hints. So it is OK to use the original one. I'll delete it later.

File: src/services/scene_manager.py (hunk 1 above)
Comment (JimZhouZZY) [reply src/services/scene_manager.py:14]: Nice catch. It is not required. I used to change the path of maps but regreted. This line is a left over.

File: src/game_entities/foe.py (hunk 5 above)
Comment (JimZhouZZY) [reply src/game_entities/foe.py:144]: To seperate two keywords. For example "Undead, Large". It's the original method, but I forgot to add ` + ", "` at the end of Line144.

File: main.py (hunk 2 above)
Comment (JimZhouZZY) [reply main.py:34]: Understood, modified

File: src/gui/language.py
Comment (JimZhouZZY) [reply src/gui/language.py:1]: Good idea.

File: maps/level_1/map.tmx (hunk 3 above)
Comment (JimZhouZZY) [reply maps/level_1/map.tmx:3]: Keep the properties part in map.tmx does not harm. We just need to copy this part to map_properties.tmx and translate map_properties.tmx.

File: src/gui/info_box.py
Comment (JimZhouZZY) [reply src/gui/info_box.py:1]: Made change in pygamepopup. The game now need new version of pygamepopup to run.

**Comment by JimZhouZZY** [comment]
The game after deleting info_box.py requires new release of pygamepopup to run.

File: src/gui/info_box.py
Comment (Grimmys) [reply src/gui/info_box.py:1]: I totally forgot to read the description coming with your PR, I'm sorry.
Going to do it.
Thanks for opening the PR in the other repo. 🙂

Comment (Grimmys) [reply src/gui/info_box.py:1]: New version of pygamepopup has been released: https://pypi.org/project/pygame-popup/
You can update the required version of the package in `requirements.txt` file.

File: main.py (hunk 2 above)
Comment (Grimmys) [reply main.py]: Looks way better!

File: src/game_entities/entity.py (hunk 4 above)
Comment (JimZhouZZY) [reply src/game_entities/entity.py]: Done.

File: data/en/text.py [hunk 6]
```diff
@@ -249,95 +249,102 @@ def f_RECOVER_NUMBER_HP(power):
...
+        "barding_magenta": "Barding Magenta",
+        "brown_boots": "Brown Boots",
+        "black_boots": "Black Boots",
+        "gold_boots": "Gold Boots",
+        "wooden_shield": "Wooden Shield",
+        "pocket_knife": "Pocket Knife",
+        "basic_spear": "Basic Spear",
+        "basic_halberd": "Basic Halberd",
+    },
+    "Effects&Alternations": {
```
Comment (Grimmys) [inline data/en/text.py:299]: I'm sorry but I feel like this naming format is wrong.

It should not be PascalCase, but snake_case.

So `items`, `foe_keywords` etc.

Also, I think effects and alt**era**tions should be splited into different categories, even if they are close. They are close, but should not be confused imo.
Same for races and classes.

**Review by Grimmys — CHANGES_REQUESTED** [review]

File: data/en/text.py (hunk 6 above)
Comment (JimZhouZZY) [reply data/en/text.py:299]: >It should not be PascalCase, but snake_case.

Okay. Ill rename it.

>Also, I think effects and alterations should be splited into different categories, even if they are close. They are close, but should not be confused imo.

I merged these items because the names seemed to be mixed up in some tests (class was treated as race). Modifying it might introduce some duplication, but I'm at ease with your words. And thanks for pointing out the "alterations" , bruh.

Comment (JimZhouZZY) [reply data/en/text.py:299]: I think races and classes should still be merged. because
* races and classes are mixed in some places in the project such as
equipment.py Line104
```python
            if key in ("classes", "races"):
                formatted_string += (
                    ", ".join(
                        [
                            TRANSLATIONS["Races&Classes"][restriction.lower().replace(" ", "_")]
                            for restriction in self.restrictions[key]
                        ]
                    )
                    + ", "
                )
```
Seperating them need new refactor on those mixed parts
* races and classes have a low change of being duplicated

Comment (Grimmys) [reply data/en/text.py]: I see, yes it is introducing more refactor to split them then let's not split, at least for now.

**Comment by Grimmys** [comment]
LGTM, thanks a lot for this big feature 🚀

**Review by Grimmys — APPROVED** [review]
